### PR TITLE
ensure consistent resource promise

### PR DESCRIPTION
**Pull request by bholloway**
Best reviewed commit at a time.

Primarily I want to ensure that the resource `promise` will always resolve `data` or reject `error`. The main culprit being `update()` where the `promise` becomes immediately mismatched with the `data`.

Timeouts should also be rejecting errors IMO but are not included here. The logic to do that is not really more complicated  but there is a more general problem that `expiresAt` and `promise` need to be set during hydration rather than being mutated on first access. I attempted this refactor but ultimately got stuck on the tests.

Additionally in the remote resource fetch
* I made the the different async results more declarative for better clarity
* covered the edge case where an explicit update aborts in-flight fetch.

I did a little work to make the tests more consistent but there is still a lot that needs doing there.

### CODE REVIEW

**Review by jackbrown — COMMENTED** [review]
@bholloway does this fix #104?

**Comment by bholloway** [comment]
@jackbrown not specifically.

This is to fix a conceptual correctness problem - IMO the `promise` should resolve `data` or `error` and never stale `data`. Specifically `update(() => null)` to delete an item should not keep a reference to data in the promise.

In react the `promise` resolved value should not really be used, there are better ways for observing resource `data`. However in playing with some "dependent resources" implementation outside react I find the promise is needed as an indicator of future value.

Since I started the branch I think this correctness with the `promise` was partly addressed during the LRU changes so ultimately there was not a lot to do. Although I was tempted to make other fixes to tests etc which would be good to land. All that is to say we should discuss offline when/how these changes should land.

File: src/controllers/resource-store/index.tsx [hunk 1]
```diff
@@ -49,17 +49,20 @@ export const actions: Actions = {
...
+    const data = getNewSliceData(slice.data);
 
     dispatch(
-      setResourceState(type, key, {
-        ...slice,
-        data: getNewSliceData(slice.data),
-        expiresAt: getExpiresAt(maxAge),
-        accessedAt: getAccessedAt(),
-      })
+      data == null
```
Comment (liamqma) [inline src/controllers/resource-store/index.tsx:56]: We are currently working on treating `null` as a valid data. Sorry for making things complicated but I am unsure that resource is removable if data is null.

Comment (bholloway) [reply src/controllers/resource-store/index.tsx:56]: In that case it's easy to not delete the record here and just treat `null` as any other update. Either way the `promise` should resolve the `null` for consistency, which is the purpose of my PR. That said I think there needs to be some seperate semantics for a `delete()` operation.

However the bigger problem I see with `null` being valid data is that it is indistinguishable from the initial state. That is a big problem IMO for determining what state the resource is in.

Comment (liamqma) [reply src/controllers/resource-store/index.tsx:56]: > However the bigger problem I see with null being valid data is that it is indistinguishable from the initial state. That is a big problem IMO for determining what state the resource is in.

Very true. I am examining different options. One of the solutions i think of is to set initial value to be 
```
export const BASE_DEFAULT_STATE_SLICE = {
  data: "DEFAULT_STATE_DATE",
};
```
in order to be distinguishable
What do you think?

Comment (bholloway) [reply src/controllers/resource-store/index.tsx:56]: hmm not a fan, it needs to be a primitive type.

Really this is the whole point of `undefined` IMO.

Comment (bholloway) [reply src/controllers/resource-store/index.tsx:56]: I think that combined with a presence of `promise` might be enough. I would need to look at the table in the `.md` file. But certainly we need to be very precise about that table going forward.

IIRC Apollo had some `linkState`(?) which makes the state explicit. I was always a fan of that since it avoids this type of problem. However I think that such a change is a bit extreme given the current API doesn't have that.

File: src/controllers/resource-store/index.tsx [hunk 2]
```diff
@@ -104,75 +104,99 @@ export const actions: Actions = {
...
+          };
         } else {
           timeoutGuard.timerId && clearTimeout(timeoutGuard.timerId);
-          response.data = maybeData;
-          response.loading = false;
+          resolvedSlice = {
+            data: maybeData,
+            error: null, // previous error is removed
+            loading: false,
+            promise,
```
Comment (albertogasparin) [inline src/controllers/resource-store/index.tsx:157]: `pendingSlice` already has `promise` being set. Is there any value setting it to `resolvedSlice` ?

File: src/controllers/resource-store/index.tsx [hunk 3]
```diff
@@ -104,75 +104,99 @@ export const actions: Actions = {
...
+        error,
+        loading: false,
+      };
     }
 
-    response.expiresAt = getExpiresAt(
-      options.prefetch && maxAge < PREFETCH_MAX_AGE ? PREFETCH_MAX_AGE : maxAge
-    );
+    // do not overwrite a slice that was updated during async
+    const laterSlice = getSliceForResource(getState(), { type, key });
```
Comment (albertogasparin) [inline src/controllers/resource-store/index.tsx:178]: Are we sure about this? If I'm implementing some sort of optimistic updates, this will discard server data for client-side data that I might have manually set, and which might be incorrect (especially if server returns an error).

Comment (bholloway) [reply src/controllers/resource-store/index.tsx:178]: IMO any optimistic update needs to be done before refresh otherwise it implicitly relies on the async nature of refresh to avoid race condition.

Fundamentally if the `promise` has changed during the async we are in a difference state.

The options as I see it are
1. explicit updates do no effect the `promise`, and the `promise` mismatches `data` (current state before PR)
2. explicit updates set the `promise` consistent with `data`, but preceding async refresh will set it back when it completes
3. explicit updates set the `promise` consistent with `data`, cancels in-flight async refresh

Fixing (1) is the point of this PR so I'd say its a choice between (2) and (3).
Choosing (2) puts us in a position where the `promise` no longer reflects the future value. So is also inconsistent in its own way.
I would specifically chose (3) to force developers to consider the order. It is the only solution that makes `promise` consistent with current `data` and future async `data`.

File: src/controllers/resource-store/index.tsx (hunk 2 above)
Comment (bholloway) [reply src/controllers/resource-store/index.tsx:157]: IMO yes. In this block there is a branch where a timeout will produce `promise: null`. Without declaring the value of `promise` in each branch it is harder to compare the final state at a glance.

Comment (albertogasparin) [reply src/controllers/resource-store/index.tsx:157]: Then, should we go all explicit and no longer spread `pendingSlice` in `finalSlice` ? As I'd say either we leverage spreading for all non changing data, or we explicit set the state for every step. Seems like the latter is preferable for clarity, so I'd enforce it

File: src/controllers/resource-store/index.tsx (hunk 3 above)
Comment (albertogasparin) [reply src/controllers/resource-store/index.tsx:178]: > IMO any optimistic update needs to be done before refresh otherwise it implicitly relies on the async nature of refresh to avoid race condition.

Agree. That is why I would make the async flow win anyway.

> Fundamentally if the promise has changed during the async we are in a difference state.

I'm confused on when this might happen. If we exclude a manual update, then there should be no way for something else to change that promise that is not this async flow itself. 
If this is a way to mitigate #104, then I think we need to fix that bug and not account for that eventuality at all.

**Comment by bholloway** [comment]
I'll drop this because of the conflict with the browser-only change.

Good discussion about race conditions with refresh. I will raise a separate PR with a more refined approach.
